doc/developer/platform/formalism.md

```diff
@@ -13,57 +13,370 @@ It does not intend to describe the **current behavior** of the system.
+all times `t` such that `t` is later than or equal to `since`, and `t` is *not*
+later than or equal to `upper`, reads of `ptvc` at `t` are correct: `read(ptvc,
+t) = read(tvc, t)`. In the context of a pTVC, we say that such a time `t`
+is *between* `since` and `upper`.
+
+![A diagram showing past, readable, and future times in a
+pTVC.](assets/since-readable-upper.jpeg)
+
+In this diagram, times are shown as a two-dimensional lattice forming a partial
+order; time generally flows left to right, and the precise order is shown by
+arrows. The `since` frontier is formed by two times labeled "since"; the
+`upper` frontier is formed by two times labeled "upper". Times later than
+`upper` are drawn in orange: these times are in the unknown future of the pTVC.
+Times not later than or equal to `since` are drawn in gray: these times are in
+the forgotten past. Only times in blue--those between `since` and `upper`--can
+be correctly read.
+
+## Update-set representations of TVCs and pTVCs
+
+Because the set of times is in general very large, and because TVCs usually
+change incrementally, it is usually more efficient to represent a TVC as a set
+of *updates* which encode changes between collection versions. An *update* is a
+triple `[data, time, diff]`. The `data` specifies what element of the
+collection version is changing, the `time` gives the time the change took
+place, and the `diff` is the change in multiplicity for that element.
+
+Given a collection version `cv1` and an update `u = [data, time, diff]`, we can
+*apply* `u` to `cv1` to produce a new collection version `cv2`: `apply(cv1, u)
+=> cv2`. `cv2` is exactly `cv1`, but with `multiplicity(cv2, data) =
+multiplicity(cv1, data) + diff`.
+
+We can also apply a countable set of updates to a collection version by
+applying each update in turn. We write this `apply(collection, updates)`. Since
+`apply` is commutative, the order does not matter. In practice, these are
+computers: all sets are countable. (If you're wondering about uncountable sets,
+the [Möbius
+inversion](https://en.wikipedia.org/wiki/M%C3%B6bius_inversion_formula) might
+be helpful.)
+
+Given a time `t` and two collection versions `cv1` and `cv2`, we can find their
+*difference*: a minimal set of updates which, when applied to `cv1`, yields
+`cv2`.
+
+```
+diff(t, cv1, cv2) = { u = [data, t, diff] |
+                      (data in cv1 or data in cv2) and
+                      (diff = multiplicity(cv2, data) - multiplicity(cv1, data))
+                    }
+```
+
+Given a frontier `f`, a set of updates `updates` can be *compacted* together
+into a (likely smaller) set of updates where the sum of the diffs for each
+`data` in the set is preserved, and every update has a time `t` such that `f
+<=t t`. We write this as `compact(updates, f)`. Since the diffs for each data
+are unchanged, and updates commute, `apply(c, updates) = apply(c,
+compact(updates, f))` for any collection version `c`.
```

Comment:
I missed this, and we have talked since then, so possibly we should resync! If the intent is something like "copy/paste in the compaction lemmas / intuition" that works for me with some editing. If it is trying to get at a different point, I missed it!